File: apps/desktop/electron/db/sqlite.ts
import Database from 'better-sqlite3'
import { DatabaseConnection, ConnectionConfig, StorageAnalysis, QueryResult, SchemaInfo } from './index'
import { copyFileSync, existsSync, statSync } from 'fs'

export class SQLiteConnection extends DatabaseConnection {
  private db?: Database.Database

  async connect(): Promise<void> {
    if (!this.config.filePath || !existsSync(this.config.filePath)) {
      throw new Error('SQLite file path is required and must exist')
    }

    this.db = new Database(this.config.filePath)
    this.connected = true
  }

  async disconnect(): Promise<void> {
    if (this.db) {
      this.db.close()
      this.db = undefined
      this.connected = false
    }
  }

  async testConnection(): Promise<void> {
    if (!this.db) {
      await this.connect()
    }
    
    const result = this.db!.prepare('SELECT 1 as test').get()
    if (!result) {
      throw new Error('Connection test failed')
    }
  }

  async analyzeStorage(): Promise<StorageAnalysis> {
    if (!this.db) {
      await this.connect()
    }

    const db = this.db!

    // Get database file size
    const stats = statSync(this.config.filePath!)
    const totalSize = stats.size

    // Get table information
    const tablesResult = db.prepare(`
      SELECT 
        name,
        (SELECT COUNT(*) FROM sqlite_master WHERE type = 'index' AND tbl_name = t.name) as index_count
      FROM sqlite_master t
      WHERE type = 'table' AND name NOT LIKE 'sqlite_%'
      ORDER BY name
    `).all()

    const tables = []
    let totalTableSize = 0

    for (const tableRow of tablesResult) {
      const tableName = tableRow.name
      
      // Get table size and row count
      const sizeResult = db.prepare(`
        SELECT 
          COUNT(*) as row_count,
          SUM(pgsize) as page_size
        FROM pragma_page_count(),
        pragma_page_size(),
        (SELECT COUNT(*) FROM ${tableName}) as row_count
      `).get()

      const rowCount = sizeResult?.row_count || 0
      const pageSize = sizeResult?.page_size || 0
      const tableSize = pageSize

      totalTableSize += tableSize

      tables.push({
        name: tableName,
        size: tableSize,
        rowCount,
        indexSize: 0, // SQLite doesn't easily provide index size per table
        bloat: 0
      })
    }

    // Get index information
    const indexesResult = db.prepare(`
      SELECT 
        name,
        tbl_name as table_name
      FROM sqlite_master 
      WHERE type = 'index' AND name NOT LIKE 'sqlite_%'
      ORDER BY name
    `).all()

    const indexes = indexesResult.map(row => ({
      name: row.name,
      tableName: row.table_name,
      size: 0, // SQLite doesn't easily provide per-index size
      bloat: 0
    }))

    const largestTable = tables.length > 0 ? tables[0] : { name: 'N/A', size: 0, rowCount: 0 }

    return {
      totalSize,
      tableCount: tables.length,
      indexCount: indexes.length,
      largestTable,
      tables,
      indexes,
      lastAnalyzed: new Date().toISOString()
    }
  }

  async executeQuery(query: string): Promise<QueryResult> {
    if (!this.db) {
      await this.connect()
    }

    const db = this.db!
    const startTime = Date.now()

    try {
      const stmt = db.prepare(query)
      
      if (query.trim().toLowerCase().startsWith('select')) {
        const rows = stmt.all()
        const executionTime = Date.now() - startTime

        // Get column names from the first row
        const columns = rows.length > 0 ? Object.keys(rows[0]) : []

        // Get explain plan for SELECT queries
        let explainPlan
        try {
          const explainResult = db.prepare(`EXPLAIN QUERY PLAN ${query}`).all()
          explainPlan = explainResult
        } catch (error) {
          // Ignore explain errors
        }

        return {
          columns,
          rows,
          rowCount: rows.length,
          executionTime,
          explainPlan
        }
      } else {
        const result = stmt.run()
        const executionTime = Date.now() - startTime

        return {
          columns: [],
          rows: [],
          rowCount: result.changes || 0,
          executionTime,
          explainPlan: undefined
        }
      }
    } catch (error) {
      throw new Error(`Query execution failed: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }

  async getSchema(): Promise<SchemaInfo> {
    if (!this.db) {
      await this.connect()
    }

    const db = this.db!

    // Get tables
    const tablesResult = db.prepare(`
      SELECT name 
      FROM sqlite_master 
      WHERE type = 'table' AND name NOT LIKE 'sqlite_%'
      ORDER BY name
    `).all()

    const tables = []

    for (const tableRow of tablesResult) {
      const tableName = tableRow.name

      // Get columns using PRAGMA table_info
      const columnsResult = db.prepare(`PRAGMA table_info(${tableName})`).all()

      const columns = columnsResult.map(col => ({
        name: col.name,
        type: col.type,
        nullable: col.notnull === 0,
        defaultValue: col.dflt_value
      }))

      // Get indexes
      const indexesResult = db.prepare(`
        SELECT 
          name,
          sql
        FROM sqlite_master 
        WHERE type = 'index' AND tbl_name = ? AND name NOT LIKE 'sqlite_%'
        ORDER BY name
      `).all(tableName)

      const indexes = indexesResult.map(idx => ({
        name: idx.name,
        columns: [], // Would need to parse SQL to extract columns
        unique: idx.sql?.includes('UNIQUE') || false
      }))

      tables.push({
        name: tableName,
        columns,
        indexes
      })
    }

    return {
      schemas: [{
        name: 'main',
        tables
      }]
    }
  }

  async createBackup(backupPath: string): Promise<{ path: string; size: number }> {
    if (!this.db) {
      await this.connect()
    }

    // For SQLite, we can simply copy the file
    try {
      copyFileSync(this.config.filePath!, backupPath)
      const stats = statSync(backupPath)
      return { path: backupPath, size: stats.size }
    } catch (error) {
      throw new Error(`Failed to create SQLite backup: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }

  async restoreBackup(backupPath: string): Promise<void> {
    if (!this.db) {
      await this.connect()
    }

    // For SQLite restore, we need to close the current connection first
    if (this.db) {
      this.db.close()
      this.connected = false
    }

    try {
      // Copy the backup file to the original location
      copyFileSync(backupPath, this.config.filePath!)
      
      // Reconnect
      this.db = new Database(this.config.filePath!)
      this.connected = true
    } catch (error) {
      throw new Error(`Failed to restore SQLite backup: ${error instanceof Error ? error.message : 'Unknown error'}`)
    }
  }
}

File: db_storage_manager/db/sqlite.py
"""
SQLite database connection
"""

import asyncio
import shutil
from pathlib import Path
from typing import Any, Dict, List
import aiosqlite

from .base import (
    DatabaseConnection,
    ConnectionConfig,
    StorageAnalysis,
    QueryResult,
    SchemaInfo,
)


class SQLiteConnection(DatabaseConnection):
    """SQLite database connection"""

    def __init__(self, config: ConnectionConfig):
        super().__init__(config)
        self.connection = None

    async def connect(self) -> None:
        """Connect to SQLite database"""
        db_path = self.config.database or self.config.connection_string
        if not db_path:
            raise ValueError("SQLite database path is required")

        self.connection = await aiosqlite.connect(db_path)
        self.connected = True

    async def disconnect(self) -> None:
        """Disconnect from SQLite database"""
        if self.connection:
            await self.connection.close()
            self.connection = None
        self.connected = False

    async def analyze_storage(self) -> StorageAnalysis:
        """Analyze SQLite storage"""
        if not self.connection:
            await self.connect()

        # Get tables
        cursor = await self.connection.execute(
            """
            SELECT name
            FROM sqlite_master
            WHERE type = 'table' AND name NOT LIKE 'sqlite_%'
            ORDER BY name
        """
        )

        tables = []
        total_size = 0
        db_path = Path(self.config.database or self.config.connection_string)
        db_size = db_path.stat().st_size if db_path.exists() else 0

        table_rows = await cursor.fetchall()
        table_count = len(table_rows)

        for (table_name,) in table_rows:
            # Get row count
            count_cursor = await self.connection.execute(
                f'SELECT COUNT(*) FROM "{table_name}"'
            )
            row_count = (await count_cursor.fetchone())[0]

            # Approximate table size (distribute total DB size)
            table_size = db_size // table_count if table_count > 0 else 0

            tables.append(
                {
                    "name": table_name,
                    "size": table_size,
                    "rowCount": row_count,
                    "indexSize": 0,
                    "bloat": 0.0,
                }
            )
            total_size += table_size

        # Get indexes
        cursor = await self.connection.execute(
            """
            SELECT name, tbl_name
            FROM sqlite_master
            WHERE type = 'index' AND name NOT LIKE 'sqlite_%'
            ORDER BY name
        """
        )

        indexes = []
        for index_name, table_name in await cursor.fetchall():
            indexes.append(
                {
                    "name": index_name,
                    "tableName": table_name,
                    "size": 0,
                    "bloat": 0.0,
                }
            )

        largest_table = (
            tables[0] if tables else {"name": "N/A", "size": 0, "rowCount": 0}
        )

        return {
            "totalSize": total_size,
            "tableCount": len(tables),
            "indexCount": len(indexes),
            "largestTable": largest_table,
            "tables": tables,
            "indexes": indexes,
            "lastAnalyzed": __import__("datetime").datetime.now().isoformat(),
        }

    async def execute_query(self, query: str, safe_mode: bool = True) -> QueryResult:
        """Execute a SQLite query"""
        if not self.connection:
            await self.connect()

        if safe_mode and not query.strip().upper().startswith("SELECT"):
            raise ValueError("Only SELECT queries are allowed in safe mode")

        import time

        start_time = time.time()

        try:
            cursor = await self.connection.execute(query)
            rows = await cursor.fetchall()

            if cursor.description:
                columns = [desc[0] for desc in cursor.description]
                rows_dict = [dict(zip(columns, row)) for row in rows]
            else:
                columns = ["affected_rows"]
                rows_dict = [{"affected_rows": self.connection.total_changes}]

            execution_time = int((time.time() - start_time) * 1000)

            # Get explain plan for SELECT queries
            explain_plan = None
            if query.strip().upper().startswith("SELECT"):
                try:
                    explain_cursor = await self.connection.execute(
                        f"EXPLAIN QUERY PLAN {query}"
                    )
                    explain_plan = await explain_cursor.fetchall()
                except Exception:
                    pass

            return {
                "columns": columns,
                "rows": rows_dict,
                "rowCount": len(rows_dict),
                "executionTime": execution_time,
                "explainPlan": explain_plan,
            }
        except Exception as e:
            raise RuntimeError(f"Query execution failed: {str(e)}")

    async def get_schema(self) -> SchemaInfo:
        """Get SQLite schema"""
        if not self.connection:
            await self.connect()

        # Get tables
        cursor = await self.connection.execute(
            """
            SELECT name
            FROM sqlite_master
            WHERE type = 'table' AND name NOT LIKE 'sqlite_%'
            ORDER BY name
        """
        )

        tables = []
        for (table_name,) in await cursor.fetchall():
            # Get columns using PRAGMA
            pragma_cursor = await self.connection.execute(
                f"PRAGMA table_info({table_name})"
            )
            columns_data = await pragma_cursor.fetchall()

            columns = []
            for col in columns_data:
                columns.append(
                    {
                        "name": col[1],  # column name
                        "type": col[2],  # data type
                        "nullable": col[3] == 0,  # not null flag
                        "defaultValue": col[4],  # default value
                    }
                )

            # Get indexes
            index_cursor = await self.connection.execute(
                """
                SELECT name, sql
                FROM sqlite_master
                WHERE type = 'index' AND tbl_name = ? AND name NOT LIKE 'sqlite_%'
            """,
                (table_name,),
            )

            indexes = []
            for index_name, index_sql in await index_cursor.fetchall():
                indexes.append(
                    {
                        "name": index_name,
                        "columns": [],
                        "unique": "UNIQUE" in (index_sql or "").upper(),
                    }
                )

            tables.append(
                {
                    "name": table_name,
                    "columns": columns,
                    "indexes": indexes,
                }
            )

        return {
            "schemas": [
                {
                    "name": self.config.database or "default",
                    "tables": tables,
                }
            ]
        }

    async def create_backup(self, backup_path: str) -> Dict[str, Any]:
        """Create SQLite backup"""
        if not self.connection:
            await self.connect()

        db_path = Path(self.config.database or self.config.connection_string)
        if not db_path.exists():
            raise FileNotFoundError(f"Database file not found: {db_path}")

        backup_file = Path(backup_path)
        backup_file.parent.mkdir(parents=True, exist_ok=True)

        # Copy database file
        shutil.copy2(db_path, backup_file)
        size = backup_file.stat().st_size

        return {"path": str(backup_file), "size": size}

    async def restore_backup(self, backup_path: str) -> None:
        """Restore SQLite backup"""
        backup_file = Path(backup_path)
        if not backup_file.exists():
            raise FileNotFoundError(f"Backup file not found: {backup_path}")

        db_path = Path(self.config.database or self.config.connection_string)

        # Disconnect before restoring
        await self.disconnect()

        # Copy backup file to database location
        shutil.copy2(backup_file, db_path)

        # Reconnect
        await self.connect()
